### ISSUE TITLE
JupiterTestEngine should log discovery errors instead of aborting the entire test plan

### ISSUE THREAD

**Issue by sbrannen** [junit-team/junit5#1438]
## Overview

An error reported in #1436 has made us aware that the entire test plan for JUnit Jupiter aborts whenever certain types of errors occur during the discovery phase. Consequently, a single error (e.g., an invalid class file format) can result in **zero** tests being discovered and executed.

This is, however, unacceptable behavior. The `JupiterTestEngine` should log such errors rather than abort the entire discovery process. Logging such errors allows the test plan to discover and execute as many tests as possible while still informing the user of tests that could not be properly _discovered_.

## Proposal

One possible solution would be to add try-catch blocks to `resolveClass()`, `resolveMethod()`, and `resolveUniqueId()` in `JavaElementsResolver`. These try-catch blocks should log errors and swallow them (unless they are _blacklisted_ exceptions).

## Related Issues

- #1436 

## Deliverables

- [x] Ensure that the `JupiterTestEngine` does not abort the entire discovery process when an exception is encountered.

**Comment by sbrannen** [comment]
@junit-team/junit-lambda, I'm inclined to go with the current proposal in the _Proposal_ section of this issue's description. 

Thoughts?

**Comment by sbrannen** [comment]
_in progress_

**Comment by sbrannen** [comment]
**Update**:

I'm currently tinkering with replacing the status quo:

```java
	void resolveMethod(Class<?> testClass, Method testMethod) {
		Set<TestDescriptor> potentialParents = resolveContainerWithParents(testClass);
		Set<TestDescriptor> resolvedDescriptors = resolveForAllParents(testMethod, potentialParents);

		if (resolvedDescriptors.isEmpty()) {
			logger.debug(() -> format("Method '%s' could not be resolved.", testMethod.toGenericString()));
		}

		logMultipleTestDescriptorsForSingleElement(testMethod, resolvedDescriptors);
	}
```

with the following "safe"/"logging" approach:

```java
	void resolveMethod(MethodSelector selector) {
		safelyGetJavaClass(selector).ifPresent(testClass -> {
			safelyGetJavaMethod(selector).ifPresent(testMethod -> {
				Set<TestDescriptor> potentialParents = resolveContainerWithParents(testClass);
				Set<TestDescriptor> resolvedDescriptors = resolveForAllParents(testMethod, potentialParents);

				if (resolvedDescriptors.isEmpty()) {
					logger.debug(() -> format("Method '%s' could not be resolved.", testMethod.toGenericString()));
				}

				logMultipleTestDescriptorsForSingleElement(testMethod, resolvedDescriptors);
			});

		});
	}

	private Optional<Class<?>> safelyGetJavaClass(MethodSelector selector) {
		try {
			return Optional.of(selector.getJavaClass());
		}
		catch (Exception ex) {
			logger.debug(ex, () -> format("Class '%s' could not be resolved.", selector.getClassName()));
			return Optional.empty();
		}
	}

	private Optional<Method> safelyGetJavaMethod(MethodSelector selector) {
		try {
			return Optional.of(selector.getJavaMethod());
		}
		catch (Exception ex) {
			logger.debug(ex, () -> format("Method '%s' in class '%s' could not be resolved.", selector.getMethodName(),
				selector.getClassName()));
			return Optional.empty();
		}
	}
```

@junit-team/junit-lambda, thoughts?

**Comment by sbrannen** [comment]
FYI: current work on this issue can be seen in the following branch.

https://github.com/junit-team/junit5/tree/issues/1438-safe-discovery-in-jupiter

**Comment by sbrannen** [comment]
Fixed in 3e84838c16eeef7bde8b37f049da0b0e0423e244
